core/trino-main/src/main/java/io/trino/operator/FlatHashStrategyCompiler.java

```diff
@@ -142,7 +144,53 @@ public static FlatHashStrategy compileFlatHashStrategy(List<Type> types, TypeOpe
         }
     }
 
-    private static void generateGetTotalVariableWidth(ClassDefinition definition, List<KeyField> keyFields, CallSiteBinder callSiteBinder)
+    private static ChunkClass compileFlatHashStrategyChunk(CallSiteBinder callSiteBinder, List<Type> types, TypeOperators typeOperators, int chunkNumber)
+    {
+        List<KeyField> keyFields = new ArrayList<>();
+        int fixedOffset = 0;
+        for (int i = 0; i < types.size(); i++) {
+            Type type = types.get(i);
+            keyFields.add(new KeyField(
+                    i,
+                    type,
+                    fixedOffset,
+                    fixedOffset + 1,
+                    typeOperators.getReadValueOperator(type, simpleConvention(BLOCK_BUILDER, FLAT)),
+                    typeOperators.getReadValueOperator(type, simpleConvention(FLAT_RETURN, BLOCK_POSITION_NOT_NULL)),
+                    typeOperators.getDistinctFromOperator(type, simpleConvention(FAIL_ON_NULL, FLAT, BLOCK_POSITION_NOT_NULL)),
+                    typeOperators.getHashCodeOperator(type, simpleConvention(FAIL_ON_NULL, FLAT)),
+                    typeOperators.getHashCodeOperator(type, simpleConvention(FAIL_ON_NULL, BLOCK_POSITION_NOT_NULL))));
+            fixedOffset += 1 + type.getFlatFixedSize();
+        }
+
+        ClassDefinition definition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                makeClassName("FlatHashStrategy$" + chunkNumber),
+                type(Object.class),
+                type(FlatHashStrategy.class));
+
+        definition.declareDefaultConstructor(a(PRIVATE));
+
+        MethodDefinition getTotalVariableWidthChunk = generateGetTotalVariableWidthChunk(definition, keyFields, callSiteBinder);
+        MethodDefinition readFlatChunk = generateReadFlatChunk(definition, keyFields, callSiteBinder);
```

Comment:
It probably should keep track of channel index, e.g: `0, 500, 1000`, so that correct offsets are used when accessing: `byte[] fixedChunk` and `byte[] variableChunk` and `BlockBuilder[] blockBuilders);`

Reply:
I assume there is a bug like this somewhere in here, but my memory is that all of the indexes come from KeyField and not the oridinal of the key field object in the list… If you look at the code, I don’t think we are using an indexed for loop anywhere.